flax/linen/linear.py

```diff
@@ -51,6 +54,38 @@ def _canonicalize_tuple(x: Union[Sequence[int], int]) -> Tuple[int, ...]:
     return (x,)
 
 
+def _canonicalize_dtypes(
+    input_dtype: InexactDType,
+    param_dtype: Optional[InexactDType],
+    computation_dtype: Optional[InexactDType]) -> Tuple[InexactDType,
+                                                        InexactDType]:
+  returned_param_dtype = input_dtype if param_dtype is None else param_dtype
+  dtype = (jnp.result_type(input_dtype, returned_param_dtype)
+           if computation_dtype is None else computation_dtype)
+
+  assert jnp.issubdtype(input_dtype, jnp.inexact)
+  if jnp.issubdtype(input_dtype, jnp.complexfloating):
+    assert jnp.issubdtype(returned_param_dtype, jnp.complexfloating)
+    assert jnp.issubdtype(dtype, jnp.complexfloating)
+  return returned_param_dtype, dtype
+
+
+def _canonicalize_numeric_dtypes(
+    input_dtype: NumericDType,
+    param_dtype: Optional[NumericDType],
+    computation_dtype: Optional[NumericDType]) -> Tuple[NumericDType,
+                                                        NumericDType]:
+  returned_param_dtype = input_dtype if param_dtype is None else param_dtype
+  dtype = (jnp.result_type(input_dtype, returned_param_dtype)
+           if computation_dtype is None else computation_dtype)
+
+  assert jnp.issubdtype(input_dtype, jnp.number)
+  if jnp.issubdtype(input_dtype, jnp.complexfloating):
+    assert jnp.issubdtype(returned_param_dtype, jnp.complexfloating)
+    assert jnp.issubdtype(dtype, jnp.complexfloating)
```

Comment:
casting back to real is currently allowed. We should look into disabling this in JAX instead of doing it inconsistently in this PR I think

Reply:
Oh, good point!  For some reason I thought it was a narrowing error.

I agree that we shouldn't try to fix this here.  Is there a Jax or numpy issue somewhere that you know of?  Maybe we should at least create one.